[Robustness] Containerized ecchronos restarting tomcat when Cassandra peer is overloaded

Story Description:

1. Setup
* CIL Datacenter with 28 pods
* Traffic simulator producing enough traffic "on the verge"of overload, so that if one pod is down, the system starts to reject requests due to ECOP inbound

2. Step
* Provoke a complete restart of the datacenter with the following command:
kubectl get pod | grep rack | awk '{ print $1 }' | xargs kubectl delete pod

3. Expected Result
* I expect that the Tomcat server that is running ecchronos does not get killed, restarting the ecchronos container.
* I expect that while the pods are starting up and the cassandra container from the pods are starting up and getting overloaded, this does not make ecchronos to get killed.
* In the end, I expect that the pods are able to get up and not be locked at startup

Acceptance Criteria:

* Ecchronos does not stop/break when cassandra is overloaded
* The target DC doesn't get stuck in a complete restart

Definition of Done:

Repeating the same test case from the "Story Description" and the ecchronos containers do not restart during the attempts

Log Excerpt:

Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.ericsson.bss.cassandra.ecchronos.application.spring.RepairHistoryBean$$EnhancerBySpringCGLIB$$ed24782f]: Constructor threw exception; nested exception is com.datastax.oss.driver.api.core.DriverTimeoutException: Query timed out after PT2S

{"version":"1.0.0","timestamp":"2024-03-19T14:35:57.949Z","severity":"warning","service_id":"eric-bss-cil-database-cd","message":"[main] [org.springframework.beans.factory.support.DisposableBeanAdapter] Invocation of close method failed on bean with name 'defaultRepairConfigurationProvider': java.lang.NullPointerException"}
{"version":"1.0.0","timestamp":"2024-03-19T14:35:57.952Z","severity":"info","service_id":"eric-bss-cil-database-cd","message":"[main] [org.apache.catalina.core.StandardService] Stopping service [Tomcat]"}

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,7 @@
 
 ## Version 5.0.1
 
+* Containerized ecchronos restarting tomcat when Cassandra peer is overloaded - Issue #650
 * Improve hang preventing task - Issue #544
 * Improve Description of unwind_ratio - Issue #628
 

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
@@ -30,6 +30,7 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.ericsson.bss.cassandra.ecchronos.application.config.Config;
+import com.ericsson.bss.cassandra.ecchronos.application.config.RetryPolicy;
 import com.ericsson.bss.cassandra.ecchronos.application.config.security.Security;
 import com.ericsson.bss.cassandra.ecchronos.connection.NativeConnectionProvider;
 import com.ericsson.bss.cassandra.ecchronos.connection.impl.LocalNativeConnectionProvider;
@@ -81,7 +82,8 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
                 .withNodeStateListener(defaultRepairConfigurationProvider);
 
         myLocalNativeConnectionProvider = establishConnection(nativeConnectionBuilder,
-                host, port, nativeConfig.getTimeout().getConnectionTimeout(TimeUnit.MILLISECONDS));
+                host, port, nativeConfig.getTimeout().getConnectionTimeout(TimeUnit.MILLISECONDS),
+                nativeConfig.getRetryPolicy());
     }
 
     public DefaultNativeConnectionProvider(final Config config,
@@ -98,11 +100,10 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
             final LocalNativeConnectionProvider.Builder builder,
             final String host,
             final int port,
-            final long timeout)
+            final long timeout,
+            final RetryPolicy retryPolicy)
     {
-        long startTime = System.currentTimeMillis();
-        long endTime = startTime + timeout;
-        while (endTime > System.currentTimeMillis())
+        for (int count = 1; count <= retryPolicy.getMaxAttempts(); count++)
         {
             try
             {
@@ -110,19 +111,33 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
             }
             catch (AllNodesFailedException | IllegalStateException e)
             {
+                LOG.warn("Unable to connect through CQL using {}:{}, retrying", host, port);
+                long delay = retryPolicy.currentDelay(count);
+                long currentTime = System.currentTimeMillis();
+                long endTime = currentTime + timeout;
                 try
                 {
-                    LOG.warn("Unable to connect through CQL using {}:{}, retrying", host, port);
-                    if (timeout != 0)
+                    if (currentTime + delay > endTime)
                     {
-                        LOG.debug("Connection failed, retrying in 5 seconds", e);
+                        String msg = String.format(
+                                "Connection failed in attempt %d of %d. Retrying in 5 seconds.",
+                                count, retryPolicy.getMaxAttempts());
+                        LOG.warn(msg);
                         Thread.sleep(SLEEP_TIME);
+                    }
+                    else
+                    {
+                        String msg = String.format(
+                                "Connection failed in attempt %d of %d. Retrying in %d seconds.",
+                                count, retryPolicy.getMaxAttempts(), TimeUnit.MILLISECONDS.toSeconds(delay));
+                        LOG.warn(msg);
+                        Thread.sleep(delay);
                     }
                 }
                 catch (InterruptedException e1)
                 {
                     LOG.error("Unexpected interrupt while trying to connect to Cassandra", e1);
-                    throw e;
+                    throw new RuntimeException(e1);
                 }
             }
         }

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
@@ -97,51 +97,62 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
     }
 
     private static LocalNativeConnectionProvider establishConnection(
-            final LocalNativeConnectionProvider.Builder builder,
-            final String host,
-            final int port,
-            final long timeout,
-            final RetryPolicy retryPolicy)
+        final LocalNativeConnectionProvider.Builder builder,
+        final String host,
+        final int port,
+        final long timeout,
+        final RetryPolicy retryPolicy)
     {
-        for (int count = 1; count <= retryPolicy.getMaxAttempts(); count++)
+        for (int attempt = 1; attempt <= retryPolicy.getMaxAttempts(); attempt++)
         {
             try
             {
-                return builder.build();
+                return tryEstablishConnection(builder);
             }
             catch (AllNodesFailedException | IllegalStateException e)
             {
-                LOG.warn("Unable to connect through CQL using {}:{}, retrying", host, port);
-                long delay = retryPolicy.currentDelay(count);
-                long currentTime = System.currentTimeMillis();
-                long endTime = currentTime + timeout;
-                try
-                {
-                    if (currentTime + delay > endTime)
-                    {
-                        String msg = String.format(
-                                "Connection failed in attempt %d of %d. Retrying in 5 seconds.",
-                                count, retryPolicy.getMaxAttempts());
-                        LOG.warn(msg);
-                        Thread.sleep(SLEEP_TIME);
-                    }
-                    else
-                    {
-                        String msg = String.format(
-                                "Connection failed in attempt %d of %d. Retrying in %d seconds.",
-                                count, retryPolicy.getMaxAttempts(), TimeUnit.MILLISECONDS.toSeconds(delay));
-                        LOG.warn(msg);
-                        Thread.sleep(delay);
-                    }
-                }
-                catch (InterruptedException e1)
-                {
-                    LOG.error("Unexpected interrupt while trying to connect to Cassandra", e1);
-                    throw new RuntimeException(e1);
-                }
+                handleRetry(attempt, retryPolicy, host, port);
             }
         }
-        return builder.build();
+        throw new RuntimeException("Failed to establish connection after all retry attempts.");
+    }
+
+    private static LocalNativeConnectionProvider tryEstablishConnection(
+        final LocalNativeConnectionProvider.Builder builder)
+    {
+        try
+        {
+            return builder.build();
+        }
+        catch (AllNodesFailedException | IllegalStateException e)
+        {
+            throw e;
+        }
+    }
+
+    private static void handleRetry(
+        final int attempt,
+        final RetryPolicy retryPolicy,
+        final String host,
+        final int port)
+    {
+        LOG.warn("Unable to connect through CQL using {}:{}, retrying.", host, port);
+        long delay = retryPolicy.currentDelay(attempt);
+        String msg = String.format(
+                "Connection failed in attempt %d of %d. Retrying in %d seconds.",
+                attempt, retryPolicy.getMaxAttempts(), TimeUnit.MILLISECONDS.toSeconds(delay));
+        LOG.warn(msg);
+        try
+        {
+            Thread.sleep(delay);
+        }
+        catch (InterruptedException e1)
+        {
+            LOG.error(
+                "InterruptedException caught during the delay time, while trying to reconnect to Cassandra. Reason: {}",
+                e1);
+            throw new RuntimeException(e1);
+        }
     }
 
     @Override

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
@@ -23,6 +23,7 @@ import com.datastax.oss.driver.api.core.auth.AuthProvider;
 import com.datastax.oss.driver.api.core.metadata.Node;
 import com.datastax.oss.driver.api.core.ssl.SslEngineFactory;
 import com.ericsson.bss.cassandra.ecchronos.application.config.connection.NativeConnection;
+import com.ericsson.bss.cassandra.ecchronos.application.config.exceptions.RetryPolicyException;
 import com.ericsson.bss.cassandra.ecchronos.connection.CertificateHandler;
 import com.ericsson.bss.cassandra.ecchronos.core.repair.DefaultRepairConfigurationProvider;
 import io.micrometer.core.instrument.MeterRegistry;
@@ -38,8 +39,6 @@ import com.ericsson.bss.cassandra.ecchronos.connection.impl.LocalNativeConnectio
 public class DefaultNativeConnectionProvider implements NativeConnectionProvider
 {
     private static final Logger LOG = LoggerFactory.getLogger(DefaultNativeConnectionProvider.class);
-
-    private static final int SLEEP_TIME = 5000;
 
     private final LocalNativeConnectionProvider myLocalNativeConnectionProvider;
 
@@ -111,10 +110,10 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
             }
             catch (AllNodesFailedException | IllegalStateException e)
             {
-                handleRetry(attempt, retryPolicy, host, port);
+                handleRetry(attempt, retryPolicy, host, port, timeout);
             }
         }
-        throw new RuntimeException("Failed to establish connection after all retry attempts.");
+        throw new RetryPolicyException("Failed to establish connection after all retry attempts.");
     }
 
     private static LocalNativeConnectionProvider tryEstablishConnection(
@@ -126,6 +125,7 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
         }
         catch (AllNodesFailedException | IllegalStateException e)
         {
+            LOG.error("Unexpected interrupt while trying to connect to Cassandra", e);
             throw e;
         }
     }
@@ -134,10 +134,17 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
         final int attempt,
         final RetryPolicy retryPolicy,
         final String host,
-        final int port)
+        final int port,
+        final long timeout)
     {
         LOG.warn("Unable to connect through CQL using {}:{}, retrying.", host, port);
         long delay = retryPolicy.currentDelay(attempt);
+        long currentTime = System.currentTimeMillis();
+        long endTime = currentTime + timeout;
+        if (currentTime + delay > endTime)
+        {
+            delay = timeout;
+        }
         String msg = String.format(
                 "Connection failed in attempt %d of %d. Retrying in %d seconds.",
                 attempt, retryPolicy.getMaxAttempts(), TimeUnit.MILLISECONDS.toSeconds(delay));
@@ -151,7 +158,6 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
             LOG.error(
                 "InterruptedException caught during the delay time, while trying to reconnect to Cassandra. Reason: {}",
                 e1);
-            throw new RuntimeException(e1);
         }
     }
 

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/DefaultNativeConnectionProvider.java
@@ -125,7 +125,7 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
         }
         catch (AllNodesFailedException | IllegalStateException e)
         {
-            LOG.error("Unexpected interrupt while trying to connect to Cassandra", e);
+            LOG.error("Unexpected interrupt while trying to connect to Cassandra. Reason: ", e);
             throw e;
         }
     }
@@ -145,10 +145,8 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
         {
             delay = timeout;
         }
-        String msg = String.format(
-                "Connection failed in attempt %d of %d. Retrying in %d seconds.",
-                attempt, retryPolicy.getMaxAttempts(), TimeUnit.MILLISECONDS.toSeconds(delay));
-        LOG.warn(msg);
+        LOG.warn("Connection failed in attempt {} of {}. Retrying in {} seconds.",
+        attempt, retryPolicy.getMaxAttempts(), TimeUnit.MILLISECONDS.toSeconds(delay));
         try
         {
             Thread.sleep(delay);
@@ -156,7 +154,7 @@ public class DefaultNativeConnectionProvider implements NativeConnectionProvider
         catch (InterruptedException e1)
         {
             LOG.error(
-                "InterruptedException caught during the delay time, while trying to reconnect to Cassandra. Reason: {}",
+                "InterruptedException caught during the delay time, while trying to reconnect to Cassandra. Reason: ",
                 e1);
         }
     }

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
@@ -14,7 +14,10 @@
  */
 package com.ericsson.bss.cassandra.ecchronos.application.config;
 
+import java.util.Locale;
 import java.util.concurrent.TimeUnit;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 public class RetryPolicy
@@ -25,17 +28,14 @@ public class RetryPolicy
     private static final int DISABLE_MAX_DELAY = 0;
 
     private Integer myMaxAttempts = DEFAULT_MAX_ATTEMPTS;
+    private TimeUnit myUnit = TimeUnit.SECONDS;
     private long myDelay = INITIAL_BACKOFF_INTERVAL_IN_MS;
     private long myMaxDelay = MAX_BACKOFF_INTERVAL_IN_MS;
 
-    public RetryPolicy(
-        final Integer retry,
-        final Integer delay,
-        final Integer maxDelay)
+    @JsonCreator
+    public RetryPolicy(@JsonProperty("unit") final String unit)
     {
-        myMaxAttempts = retry;
-        myDelay = TimeUnit.SECONDS.toMillis(delay);
-        myMaxDelay = TimeUnit.SECONDS.toMillis(maxDelay);
+        myUnit = TimeUnit.valueOf(unit.toUpperCase(Locale.US));
     }
 
     public RetryPolicy()
@@ -57,10 +57,14 @@ public class RetryPolicy
         return myMaxAttempts;
     }
 
-    @JsonProperty("maxAttempts")
-    public final void setMaxAttempts(final Integer maxAttempts)
+    public final long getMaxDelay()
     {
-        myMaxAttempts = maxAttempts;
+        return myMaxDelay;
+    }
+
+    public final TimeUnit getUnit()
+    {
+        return myUnit;
     }
 
     public final long getDelay()
@@ -68,20 +72,21 @@ public class RetryPolicy
         return myDelay;
     }
 
+    @JsonProperty("maxAttempts")
+    public final void setMaxAttempts(final Integer maxAttempts)
+    {
+        myMaxAttempts = maxAttempts;
+    }
+
     @JsonProperty("delay")
     public final void setDelay(final Integer delay)
     {
-        myDelay = TimeUnit.SECONDS.toMillis(delay);
-    }
-
-    public final long getMaxDelay()
-    {
-        return myMaxDelay;
+        myDelay = myUnit.toMillis(delay);
     }
 
     @JsonProperty("maxDelay")
-    public final void setMaxDelay(final Integer maxRetry)
+    public final void setMaxDelay(final Integer maxDelay)
     {
-        myMaxDelay = TimeUnit.SECONDS.toMillis(maxRetry);
+        myMaxDelay = myUnit.toMillis(maxDelay);
     }
 }

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
@@ -42,9 +42,9 @@ public class RetryPolicy
     {
     }
 
-    public final long currentDelay(final Integer count)
+    public final long currentDelay(final Integer attempt)
     {
-        long currentDelay = (long) (INITIAL_BACKOFF_INTERVAL_IN_MS * Math.pow(2, count));
+        long currentDelay = (long) (myDelay * Math.pow(2, attempt));
         if (myMaxDelay > DISABLE_MAX_DELAY & currentDelay > myMaxDelay)
         {
             currentDelay = myMaxDelay;

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
@@ -15,18 +15,18 @@
 package com.ericsson.bss.cassandra.ecchronos.application.config;
 
 import java.util.concurrent.TimeUnit;
-
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 public class RetryPolicy
 {
     private static final int DEFAULT_MAX_ATTEMPTS = 5;
-    private static final long DEFAULT_DELAY = 5000;
-    private static final long DEFAULT_MAX_DELAY = 30000;
+    private static final int INITIAL_BACKOFF_INTERVAL_IN_MS = 5000; // 5 seconds
+    private static final int MAX_BACKOFF_INTERVAL_IN_MS = 30000; // 30 seconds
+    private static final int DISABLE_MAX_DELAY = 0;
 
     private Integer myMaxAttempts = DEFAULT_MAX_ATTEMPTS;
-    private long myDelay = DEFAULT_DELAY;
-    private long myMaxDelay = DEFAULT_MAX_DELAY;
+    private long myDelay = INITIAL_BACKOFF_INTERVAL_IN_MS;
+    private long myMaxDelay = MAX_BACKOFF_INTERVAL_IN_MS;
 
     public RetryPolicy(
         final Integer retry,
@@ -40,13 +40,12 @@ public class RetryPolicy
 
     public RetryPolicy()
     {
-
     }
 
     public final long currentDelay(final Integer count)
     {
-        long currentDelay = myDelay * count;
-        if (currentDelay > myMaxDelay)
+        long currentDelay = (long) (INITIAL_BACKOFF_INTERVAL_IN_MS * Math.pow(2, count));
+        if (myMaxDelay > DISABLE_MAX_DELAY & currentDelay > myMaxDelay)
         {
             currentDelay = myMaxDelay;
         }

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
@@ -1,0 +1,88 @@
+/*
+ * Copyright 2024 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.ericsson.bss.cassandra.ecchronos.application.config;
+
+import java.util.concurrent.TimeUnit;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class RetryPolicy
+{
+    private static final int DEFAULT_MAX_ATTEMPTS = 5;
+    private static final long DEFAULT_DELAY = 5000;
+    private static final long DEFAULT_MAX_DELAY = 30000;
+
+    private Integer myMaxAttempts = DEFAULT_MAX_ATTEMPTS;
+    private long myDelay = DEFAULT_DELAY;
+    private long myMaxDelay = DEFAULT_MAX_DELAY;
+
+    public RetryPolicy(
+        final Integer retry,
+        final Integer delay,
+        final Integer maxDelay)
+    {
+        myMaxAttempts = retry;
+        myDelay = TimeUnit.SECONDS.toMillis(delay);
+        myMaxDelay = TimeUnit.SECONDS.toMillis(maxDelay);
+    }
+
+    public RetryPolicy()
+    {
+
+    }
+
+    public final long currentDelay(final Integer count)
+    {
+        long currentDelay = myDelay * count;
+        if (currentDelay > myMaxDelay)
+        {
+            currentDelay = myMaxDelay;
+        }
+        return currentDelay;
+    }
+
+    public final Integer getMaxAttempts()
+    {
+        return myMaxAttempts;
+    }
+
+    @JsonProperty("maxAttempts")
+    public final void setMaxAttempts(final Integer maxAttempts)
+    {
+        myMaxAttempts = maxAttempts;
+    }
+
+    public final long getDelay()
+    {
+        return myDelay;
+    }
+
+    @JsonProperty("delay")
+    public final void setDelay(final Integer delay)
+    {
+        myDelay = TimeUnit.SECONDS.toMillis(delay);
+    }
+
+    public final long getMaxDelay()
+    {
+        return myMaxDelay;
+    }
+
+    @JsonProperty("maxDelay")
+    public final void setMaxDelay(final Integer maxRetry)
+    {
+        myMaxDelay = TimeUnit.SECONDS.toMillis(maxRetry);
+    }
+}

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/RetryPolicy.java
@@ -17,6 +17,7 @@ package com.ericsson.bss.cassandra.ecchronos.application.config;
 import java.util.Locale;
 import java.util.concurrent.TimeUnit;
 
+import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
@@ -45,7 +46,7 @@ public class RetryPolicy
     public final long currentDelay(final Integer attempt)
     {
         long currentDelay = (long) (myDelay * Math.pow(2, attempt));
-        if (myMaxDelay > DISABLE_MAX_DELAY & currentDelay > myMaxDelay)
+        if ((myMaxDelay > DISABLE_MAX_DELAY) && (currentDelay > myMaxDelay))
         {
             currentDelay = myMaxDelay;
         }
@@ -88,5 +89,11 @@ public class RetryPolicy
     public final void setMaxDelay(final Integer maxDelay)
     {
         myMaxDelay = myUnit.toMillis(maxDelay);
+    }
+
+    @VisibleForTesting
+    public final void setUnit(final String unit)
+    {
+        myUnit = TimeUnit.valueOf(unit.toUpperCase(Locale.US));
     }
 }

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/Connection.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/Connection.java
@@ -14,6 +14,7 @@
  */
 package com.ericsson.bss.cassandra.ecchronos.application.config.connection;
 
+import com.ericsson.bss.cassandra.ecchronos.application.config.RetryPolicy;
 import com.ericsson.bss.cassandra.ecchronos.connection.CertificateHandler;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
@@ -28,6 +29,7 @@ public abstract class Connection<T>
     private Class<? extends T> myProviderClass;
     private Class<? extends CertificateHandler> myCertificateHandlerClass;
     private Timeout myTimeout = new Timeout();
+    private RetryPolicy myRetryPolicy = new RetryPolicy();
 
     @JsonProperty("host")
     public final String getHost()
@@ -57,6 +59,12 @@ public abstract class Connection<T>
     public final Timeout getTimeout()
     {
         return myTimeout;
+    }
+
+    @JsonProperty("retryPolicy")
+    public final RetryPolicy getRetryPolicy()
+    {
+        return myRetryPolicy;
     }
 
     @JsonProperty("host")
@@ -100,6 +108,12 @@ public abstract class Connection<T>
     public final void setTimeout(final Timeout timeout)
     {
         myTimeout = timeout;
+    }
+
+    @JsonProperty("retryPolicy")
+    public final void setRetryCount(final RetryPolicy retryPolicy)
+    {
+        myRetryPolicy = retryPolicy;
     }
 
     protected abstract Class<?>[] expectedConstructor();

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/exceptions/RetryPolicyException.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/exceptions/RetryPolicyException.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright 2024 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.ericsson.bss.cassandra.ecchronos.application.config.exceptions;
+
+/**
+ * An exception indicating that the connection with Cassandra
+ * Was not possible, which can means that the Cassandra is
+ * overloaded or wrong connection configuration.
+ */
+public class RetryPolicyException extends RuntimeException
+{
+    private static final long serialVersionUID = 8519513326549621415L;
+
+    public RetryPolicyException(final String message)
+    {
+        super(message);
+    }
+}

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/exceptions/package-info.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/exceptions/package-info.java
@@ -1,0 +1,19 @@
+/*
+ * Copyright 2024 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Contains exceptions related to outbound connections (CQL and JMX).
+ */
+package com.ericsson.bss.cassandra.ecchronos.application.config.exceptions;

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/BeanConfigurator.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/BeanConfigurator.java
@@ -148,7 +148,6 @@ public class BeanConfigurator
         CertificateHandler certificateHandler = createCertificateHandler(configuration, tlsSupplier);
 
         Class<?> providerClass = configuration.getConnectionConfig().getCqlConnection().getProviderClass();
-        LOG.warn("Trying to connect using " + providerClass);
 
         try
         {
@@ -165,6 +164,7 @@ public class BeanConfigurator
             }
             else
             {
+                // Check for old versions of DefaultNativeConnectionProvider
                 constructorArgs = new Object[]
                 {
                         configuration,
@@ -182,8 +182,8 @@ public class BeanConfigurator
         }
         catch (ConfigurationException ex)
         {
-            LOG.error("Error while trying to connect with Cassandra using " + providerClass, ex);
-            throw ex;
+            throw new ConfigurationException(
+                "Error while trying to connect with Cassandra using " + providerClass, ex);
         }
     }
 

--- a/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/BeanConfigurator.java
+++ b/application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/spring/BeanConfigurator.java
@@ -22,7 +22,6 @@ import java.util.function.Supplier;
 
 import com.datastax.oss.driver.api.core.CqlSession;
 import com.datastax.oss.driver.api.core.metadata.Node;
-import com.ericsson.bss.cassandra.ecchronos.application.ReloadingCertificateHandler;
 import com.ericsson.bss.cassandra.ecchronos.connection.CertificateHandler;
 import com.ericsson.bss.cassandra.ecchronos.core.repair.DefaultRepairConfigurationProvider;
 import io.micrometer.core.instrument.MeterRegistry;
@@ -35,6 +34,7 @@ import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
 import com.ericsson.bss.cassandra.ecchronos.application.ConfigurationException;
+import com.ericsson.bss.cassandra.ecchronos.application.DefaultNativeConnectionProvider;
 import com.ericsson.bss.cassandra.ecchronos.application.ReflectionUtils;
 import com.ericsson.bss.cassandra.ecchronos.application.config.Config;
 import com.ericsson.bss.cassandra.ecchronos.application.config.ConfigRefresher;
@@ -137,51 +137,99 @@ public class BeanConfigurator
                 eccCompositeMeterRegistry);
     }
 
-    private static NativeConnectionProvider getNativeConnectionProvider(
-            final Config configuration,
-            final Supplier<Security.CqlSecurity> securitySupplier,
-            final DefaultRepairConfigurationProvider defaultRepairConfigurationProvider,
-            final MeterRegistry meterRegistry) throws ConfigurationException
+    public static NativeConnectionProvider getNativeConnectionProvider(
+        final Config configuration,
+        final Supplier<Security.CqlSecurity> securitySupplier,
+        final DefaultRepairConfigurationProvider defaultRepairConfigurationProvider,
+        final MeterRegistry meterRegistry) throws ConfigurationException
     {
-        Supplier tlsSupplier = () -> securitySupplier.get().getCqlTlsConfig();
 
-        CertificateHandler certificateHandler = ReflectionUtils.construct(
-                configuration.getConnectionConfig().getCqlConnection().getCertificateHandlerClass(),
-                new Class<?>[] {
-                        Supplier.class
-                }, tlsSupplier);
+        Supplier tlsSupplier = () -> securitySupplier.get().getCqlTlsConfig();
+        CertificateHandler certificateHandler = createCertificateHandler(configuration, tlsSupplier);
+
+        Class<?> providerClass = configuration.getConnectionConfig().getCqlConnection().getProviderClass();
+        LOG.warn("Trying to connect using " + providerClass);
+
         try
         {
-            return ReflectionUtils
-                    .construct(configuration.getConnectionConfig().getCqlConnection().getProviderClass(),
-                            new Class<?>[] {
-                                    Config.class,
-                                    Supplier.class,
-                                    CertificateHandler.class,
-                                    DefaultRepairConfigurationProvider.class,
-                                    MeterRegistry.class
-                            },
-                            configuration, securitySupplier, certificateHandler, defaultRepairConfigurationProvider,
-                            meterRegistry);
+            Object[] constructorArgs;
+            if (DefaultNativeConnectionProvider.class.equals(providerClass))
+            {
+                constructorArgs = new Object[] {
+                        configuration,
+                        securitySupplier,
+                        certificateHandler,
+                        defaultRepairConfigurationProvider,
+                        meterRegistry
+                };
+            }
+            else
+            {
+                constructorArgs = new Object[]
+                {
+                        configuration,
+                        securitySupplier,
+                        defaultRepairConfigurationProvider,
+                        meterRegistry
+                };
+            }
+
+            return (NativeConnectionProvider) ReflectionUtils.construct(
+                    providerClass,
+                    getConstructorParameterTypes(providerClass),
+                    constructorArgs);
+
+        }
+        catch (ConfigurationException ex)
+        {
+            LOG.error("Error while trying to connect with Cassandra using " + providerClass, ex);
+            throw ex;
+        }
+    }
+
+    private static CertificateHandler createCertificateHandler(
+        final Config configuration,
+        final Supplier tlsSupplier) throws ConfigurationException
+    {
+        try
+        {
+            return ReflectionUtils.construct(
+                    configuration.getConnectionConfig().getCqlConnection().getCertificateHandlerClass(),
+                    new Class<?>[] {Supplier.class}, tlsSupplier);
         }
         catch (ConfigurationException e)
         {
-            if (!ReloadingCertificateHandler.class.equals(certificateHandler.getClass())
-                    && e.getCause() instanceof NoSuchMethodException)
+            if (!(e.getCause() instanceof NoSuchMethodException))
             {
-                throw new ConfigurationException("Invalid configuration, connection provider does not support"
-                        + " configured certificate handler", e);
+                throw e;
             }
+            throw new ConfigurationException("Invalid configuration, connection provider does not support"
+                    + " configured certificate handler.", e);
         }
+    }
 
-        // Check for old versions of DefaultNativeConnectionProvider
-        return ReflectionUtils
-                .construct(configuration.getConnectionConfig().getCqlConnection().getProviderClass(),
-                        new Class<?>[]{
-                                Config.class, Supplier.class, DefaultRepairConfigurationProvider.class,
-                                MeterRegistry.class
-                        },
-                        configuration, securitySupplier, defaultRepairConfigurationProvider, meterRegistry);
+    private static Class<?>[] getConstructorParameterTypes(final Class<?> providerClass)
+    {
+        if (DefaultNativeConnectionProvider.class.equals(providerClass))
+        {
+            return new Class<?>[] {
+                    Config.class,
+                    Supplier.class,
+                    CertificateHandler.class,
+                    DefaultRepairConfigurationProvider.class,
+                    MeterRegistry.class
+            };
+        }
+        else
+        {
+            return new Class<?>[]
+            {
+                    Config.class,
+                    Supplier.class,
+                    DefaultRepairConfigurationProvider.class,
+                    MeterRegistry.class
+            };
+        }
     }
 
     @Bean

--- a/application/src/main/resources/ecc.yml
+++ b/application/src/main/resources/ecc.yml
@@ -27,24 +27,25 @@ connection:
     host: localhost
     port: 9042
     ##
-    ## Connection Timeout for an CQL attempt.
+    ## Connection Timeout for a CQL attempt.
     ## Specify a time to wait for cassandra to come up.
-    ## Connection is tried based on retry policy delay calculations. Each connection attempt will use timeout to calculate CQL Connection proccess delay.
+    ## Connection is tried based on retry policy delay calculations. Each connection attempt will use the timeout to calculate CQL connection process delay.
     ##
     timeout:
       time: 60
       unit: seconds
     retryPolicy:
-      ## Max number of attempts that ecChronos will try to connect with Cassandra.
+      ## Max number of attempts ecChronos will try to connect with Cassandra.
       maxAttempts: 5
-      unit: seconds
-      ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
-      ## If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
+      ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count powered by two.
+      ## If the current attempt is 4 and the default delay is 5 seconds, so ((4(attempt) x 2) x 5(default delay)) = 40 seconds.
+      ## If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay.
       delay: 5
-      ## The max delay that one attempt can wait until run.
+      ## Maximum delay before the next connection attempt is made.
       ## Setting it as 0 will disable maxDelay and the delay interval will
       ## be calculated based on the attempt count and the default delay.
       maxDelay: 30
+      unit: seconds
     ##
     ## The class used to provide CQL connections to Apache Cassandra.
     ## The default provider will be used unless another is specified.

--- a/application/src/main/resources/ecc.yml
+++ b/application/src/main/resources/ecc.yml
@@ -27,16 +27,22 @@ connection:
     host: localhost
     port: 9042
     ##
-    ## Connection Timeout for CQL.
+    ## Connection Timeout for an CQL attempt.
     ## Specify a time to wait for cassandra to come up.
-    ## Connection is tried every five second until either the timeout time passes or the connection is successful.
+    ## Connection is tried based on retry policy delay calculations. Each connection attempt will use timeout to calculate CQL Connection proccess delay.
     ##
     timeout:
       time: 60
       unit: seconds
     retryPolicy:
+      ## Max number of attempts that ecChronos will try to connect with Cassandra.
       maxAttempts: 5
+      ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
+      ## If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
       delay: 5
+      ## The max delay that one attempt can wait until run.
+      ## Setting it as 0 will disable maxDelay and the delay interval will
+      ## be calculated based on the attempt count and the default delay.
       maxDelay: 30
     ##
     ## The class used to provide CQL connections to Apache Cassandra.

--- a/application/src/main/resources/ecc.yml
+++ b/application/src/main/resources/ecc.yml
@@ -37,6 +37,7 @@ connection:
     retryPolicy:
       ## Max number of attempts that ecChronos will try to connect with Cassandra.
       maxAttempts: 5
+      unit: seconds
       ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
       ## If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
       delay: 5

--- a/application/src/main/resources/ecc.yml
+++ b/application/src/main/resources/ecc.yml
@@ -32,8 +32,12 @@ connection:
     ## Connection is tried every five second until either the timeout time passes or the connection is successful.
     ##
     timeout:
-      time: 0
+      time: 60
       unit: seconds
+    retryPolicy:
+      maxAttempts: 5
+      delay: 5
+      maxDelay: 30
     ##
     ## The class used to provide CQL connections to Apache Cassandra.
     ## The default provider will be used unless another is specified.

--- a/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java
+++ b/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java
@@ -90,6 +90,9 @@ public class TestConfig
         assertThat(nativeConnection.getPort()).isEqualTo(9100);
         assertThat(nativeConnection.getRemoteRouting()).isFalse();
         assertThat(nativeConnection.getTimeout().getConnectionTimeout(TimeUnit.SECONDS)).isEqualTo(5);
+        assertThat(nativeConnection.getRetryPolicy().getMaxAttempts()).isEqualTo(10);
+        assertThat(nativeConnection.getRetryPolicy().getDelay()).isEqualTo(10000);
+        assertThat(nativeConnection.getRetryPolicy().getMaxDelay()).isEqualTo(35000);
         assertThat(nativeConnection.getProviderClass()).isEqualTo(TestNativeConnectionProvider.class);
         assertThat(nativeConnection.getCertificateHandlerClass()).isEqualTo(TestCertificateHandler.class);
         assertThat(nativeConnection.getDecoratorClass()).isEqualTo(TestStatementDecorator.class);
@@ -186,7 +189,10 @@ public class TestConfig
         assertThat(nativeConnection.getHost()).isEqualTo("localhost");
         assertThat(nativeConnection.getPort()).isEqualTo(9042);
         assertThat(nativeConnection.getRemoteRouting()).isTrue();
-        assertThat(nativeConnection.getTimeout().getConnectionTimeout(TimeUnit.MILLISECONDS)).isEqualTo(0);
+        assertThat(nativeConnection.getTimeout().getConnectionTimeout(TimeUnit.MILLISECONDS)).isEqualTo(60000);
+        assertThat(nativeConnection.getRetryPolicy().getMaxAttempts()).isEqualTo(5);
+        assertThat(nativeConnection.getRetryPolicy().getDelay()).isEqualTo(5000);
+        assertThat(nativeConnection.getRetryPolicy().getMaxDelay()).isEqualTo(30000);
         assertThat(nativeConnection.getProviderClass()).isEqualTo(DefaultNativeConnectionProvider.class);
         assertThat(nativeConnection.getCertificateHandlerClass()).isEqualTo(ReloadingCertificateHandler.class);
         assertThat(nativeConnection.getDecoratorClass()).isEqualTo(NoopStatementDecorator.class);
@@ -274,6 +280,9 @@ public class TestConfig
         assertThat(nativeConnection.getPort()).isEqualTo(9042);
         assertThat(nativeConnection.getRemoteRouting()).isTrue();
         assertThat(nativeConnection.getTimeout().getConnectionTimeout(TimeUnit.MILLISECONDS)).isEqualTo(0);
+        assertThat(nativeConnection.getRetryPolicy().getMaxAttempts()).isEqualTo(5);
+        assertThat(nativeConnection.getRetryPolicy().getDelay()).isEqualTo(5000);
+        assertThat(nativeConnection.getRetryPolicy().getMaxDelay()).isEqualTo(30000);
         assertThat(nativeConnection.getProviderClass()).isEqualTo(DefaultNativeConnectionProvider.class);
         assertThat(nativeConnection.getCertificateHandlerClass()).isEqualTo(ReloadingCertificateHandler.class);
         assertThat(nativeConnection.getDecoratorClass()).isEqualTo(NoopStatementDecorator.class);

--- a/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java
+++ b/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java
@@ -91,8 +91,9 @@ public class TestConfig
         assertThat(nativeConnection.getRemoteRouting()).isFalse();
         assertThat(nativeConnection.getTimeout().getConnectionTimeout(TimeUnit.SECONDS)).isEqualTo(5);
         assertThat(nativeConnection.getRetryPolicy().getMaxAttempts()).isEqualTo(10);
-        assertThat(nativeConnection.getRetryPolicy().getDelay()).isEqualTo(10000);
-        assertThat(nativeConnection.getRetryPolicy().getMaxDelay()).isEqualTo(35000);
+        assertThat(nativeConnection.getRetryPolicy().getUnit()).isEqualTo(TimeUnit.MINUTES);
+        assertThat(nativeConnection.getRetryPolicy().getDelay()).isEqualTo(600000);
+        assertThat(nativeConnection.getRetryPolicy().getMaxDelay()).isEqualTo(2100000);
         assertThat(nativeConnection.getProviderClass()).isEqualTo(TestNativeConnectionProvider.class);
         assertThat(nativeConnection.getCertificateHandlerClass()).isEqualTo(TestCertificateHandler.class);
         assertThat(nativeConnection.getDecoratorClass()).isEqualTo(TestStatementDecorator.class);
@@ -191,6 +192,7 @@ public class TestConfig
         assertThat(nativeConnection.getRemoteRouting()).isTrue();
         assertThat(nativeConnection.getTimeout().getConnectionTimeout(TimeUnit.MILLISECONDS)).isEqualTo(60000);
         assertThat(nativeConnection.getRetryPolicy().getMaxAttempts()).isEqualTo(5);
+        assertThat(nativeConnection.getRetryPolicy().getUnit()).isEqualTo(TimeUnit.SECONDS);
         assertThat(nativeConnection.getRetryPolicy().getDelay()).isEqualTo(5000);
         assertThat(nativeConnection.getRetryPolicy().getMaxDelay()).isEqualTo(30000);
         assertThat(nativeConnection.getProviderClass()).isEqualTo(DefaultNativeConnectionProvider.class);

--- a/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestRetryPolicy.java
+++ b/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestRetryPolicy.java
@@ -1,0 +1,69 @@
+package com.ericsson.bss.cassandra.ecchronos.application.config;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class TestRetryPolicy {
+
+    RetryPolicy retryPolicy = Mockito.mock(RetryPolicy.class);
+
+    @Before
+    public void setup()
+    {
+        retryPolicy.setUnit("seconds");
+        retryPolicy.setDelay(5);
+        retryPolicy.setMaxAttempts(5);
+        retryPolicy.setMaxDelay(30);
+    }
+
+    @Test
+    public void testDefaultCurrentDelay() {
+        long currentDelay = retryPolicy.currentDelay(1);
+
+        assertEquals(TimeUnit.SECONDS.toMillis(10), currentDelay);
+    }
+
+    @Test
+    public void testMaxDelay() {
+        long currentDelay = retryPolicy.currentDelay(5);
+
+        assertEquals(TimeUnit.SECONDS.toMillis(30), currentDelay);
+    }
+
+    @Test
+    public void testMaxDelayDisabled() {
+
+        retryPolicy.setMaxDelay(0);
+
+        long currentDelay = retryPolicy.currentDelay(5);
+
+        assertEquals(TimeUnit.SECONDS.toMillis(160), currentDelay);
+    }
+
+    @Test
+    public void testDefaultConfigurationInMinutes() {
+        retryPolicy.setUnit("minutes");
+        retryPolicy.setDelay(5);
+        retryPolicy.setMaxAttempts(5);
+        retryPolicy.setMaxDelay(30);
+
+        long currentDelay = retryPolicy.currentDelay(1);
+
+        assertEquals(TimeUnit.MINUTES.toMillis(10), currentDelay);
+
+        currentDelay = retryPolicy.currentDelay(5);
+
+        assertEquals(TimeUnit.MINUTES.toMillis(30), currentDelay);
+
+        retryPolicy.setMaxDelay(0);
+
+        currentDelay = retryPolicy.currentDelay(5);
+
+        assertEquals(TimeUnit.MINUTES.toMillis(160), currentDelay);
+    }
+}

--- a/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestRetryPolicy.java
+++ b/application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestRetryPolicy.java
@@ -1,3 +1,17 @@
+/*
+ * Copyright 2024 Telefonaktiebolaget LM Ericsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package com.ericsson.bss.cassandra.ecchronos.application.config;
 
 import static org.junit.Assert.assertEquals;

--- a/application/src/test/resources/all_set.yml
+++ b/application/src/test/resources/all_set.yml
@@ -20,6 +20,10 @@ connection:
     timeout:
       time: 5
       unit: seconds
+    retryPolicy:
+      maxAttempts: 10
+      delay: 10
+      maxDelay: 35
     provider: com.ericsson.bss.cassandra.ecchronos.application.config.TestConfig$TestNativeConnectionProvider
     certificateHandler: com.ericsson.bss.cassandra.ecchronos.application.config.TestConfig$TestCertificateHandler
     decoratorClass: com.ericsson.bss.cassandra.ecchronos.application.config.TestConfig$TestStatementDecorator

--- a/application/src/test/resources/all_set.yml
+++ b/application/src/test/resources/all_set.yml
@@ -24,6 +24,7 @@ connection:
       maxAttempts: 10
       delay: 10
       maxDelay: 35
+      unit: minutes
     provider: com.ericsson.bss.cassandra.ecchronos.application.config.TestConfig$TestNativeConnectionProvider
     certificateHandler: com.ericsson.bss.cassandra.ecchronos.application.config.TestConfig$TestCertificateHandler
     decoratorClass: com.ericsson.bss.cassandra.ecchronos.application.config.TestConfig$TestStatementDecorator

--- a/docs/RETRY_POLICY.md
+++ b/docs/RETRY_POLICY.md
@@ -8,22 +8,39 @@ If the calculated delay exceeds the allowed maximum (maxDelay), the maximum will
 
 ```yaml
     ##
-    ## Connection Timeout for an CQL attempt.
+    ## Connection Timeout for a CQL attempt.
     ## Specify a time to wait for cassandra to come up.
-    ## Connection is tried based on retry policy delay calculations. Each connection attempt will use timeout to calculate CQL Connection proccess delay.
+    ## Connection is tried based on retry policy delay calculations. Each connection attempt will use the timeout to calculate CQL connection process delay.
     ##
     timeout:
       time: 6
       unit: minutes
     retryPolicy:
-      ## Max number of attempts that ecChronos will try to connect with Cassandra.
-      maxAttempts: 10
-      unit: minutes
-      ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
+      ## Max number of attempts ecChronos will try to connect with Cassandra.
+      maxAttempts: 5
+      ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count powered by two.
+      ## If the current attempt is 4 and the default delay is 5 seconds, so ((4(attempt) x 2) x 5(default delay)) = 40 seconds.
       ## If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
       delay: 1
-      ## The max delay that one attempt can wait until run.
+      ## Maximum delay before the next connection attempt is made.
       ## Setting it as 0 will disable maxDelay and the delay interval will
       ## be calculated based on the attempt count and the default delay.
       maxDelay: 5
+      unit: minutes
+```
+
+With the configuration above, ecChronos will follow this procedure if it is not being able to connect with Cassandra.
+
+```log
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Unable to connect through CQL using localhost:9042, retrying.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Connection failed in attempt 1 of 5. Retrying in 120 seconds.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Unable to connect through CQL using localhost:9042, retrying.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Connection failed in attempt 2 of 5. Retrying in 240 seconds.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Unable to connect through CQL using localhost:9042, retrying.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Connection failed in attempt 3 of 5. Retrying in 300 seconds.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Unable to connect through CQL using localhost:9042, retrying.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Connection failed in attempt 4 of 5. Retrying in 300 seconds.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Unable to connect through CQL using localhost:9042, retrying.
+[main] WARN  c.e.b.c.e.a.DefaultNativeConnectionProvider - Connection failed in attempt 5 of 5. Retrying in 300 seconds.
+[main] ERROR  c.e.b.c.e.a..c.e.RetryPolicyException: Failed to establish connection after all retry attempts.
 ```

--- a/docs/RETRY_POLICY.md
+++ b/docs/RETRY_POLICY.md
@@ -1,0 +1,29 @@
+# Retry Policy
+
+The Retry Policy configures the connection retry behavior of ecChronos with Cassandra. The Timeout specifies the period of time ecChronos will wait for Cassandra to start per attempt. The retry policy sets the maximum number of connection attempts (maxAttempts), which by default is 5, specifying the time unit as seconds by deafult. Additionally, it determines the delay between each connection attempt (delay), multiplied by the current attempt count.
+
+If the calculated delay exceeds the allowed maximum (maxDelay), the maximum will be used. The maxDelay limits the maximum time a connection attempt can wait to execute. If set to 0, it disables maxDelay, and the delay interval will be calculated based on the number of attempts and the default delay. These settings aim to optimize the efficiency of connection attempts, ensuring the system waits an appropriate amount of time and makes a reasonable number of attempts before giving up.
+
+## Example on ecc.yml
+
+```yaml
+    ##
+    ## Connection Timeout for an CQL attempt.
+    ## Specify a time to wait for cassandra to come up.
+    ## Connection is tried based on retry policy delay calculations. Each connection attempt will use timeout to calculate CQL Connection proccess delay.
+    ##
+    timeout:
+      time: 6
+      unit: minutes
+    retryPolicy:
+      ## Max number of attempts that ecChronos will try to connect with Cassandra.
+      maxAttempts: 10
+      unit: minutes
+      ## Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
+      ## If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
+      delay: 1
+      ## The max delay that one attempt can wait until run.
+      ## Setting it as 0 will disable maxDelay and the delay interval will
+      ## be calculated based on the attempt count and the default delay.
+      maxDelay: 5
+```

--- a/docs/autogenerated/EccYamlFile.md
+++ b/docs/autogenerated/EccYamlFile.md
@@ -18,8 +18,12 @@
 # Connection is tried every five second until either the timeout time passes or the connection is successful.
 #
 **timeout:**
-* time: 0
+* time: 60
 * unit: seconds
+**retryPolicy:**
+* maxAttempts: 5
+* delay: 5
+* maxDelay: 30
 #
 # The class used to provide CQL connections to Apache Cassandra.
 # The default provider will be used unless another is specified.

--- a/docs/autogenerated/EccYamlFile.md
+++ b/docs/autogenerated/EccYamlFile.md
@@ -13,16 +13,22 @@
 * host: localhost
 * port: 9042
 #
-# Connection Timeout for CQL.
+# Connection Timeout for an CQL attempt.
 # Specify a time to wait for cassandra to come up.
-# Connection is tried every five second until either the timeout time passes or the connection is successful.
+# Connection is tried based on retry policy delay calculations. Each connection attempt will use timeout to calculate CQL Connection proccess delay.
 #
 **timeout:**
 * time: 60
 * unit: seconds
 **retryPolicy:**
+# Max number of attempts that ecChronos will try to connect with Cassandra.
 * maxAttempts: 5
+# Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
+# If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
 * delay: 5
+# The max delay that one attempt can wait until run.
+# Setting it as 0 will disable maxDelay and the delay interval will
+# be calculated based on the attempt count and the default delay.
 * maxDelay: 30
 #
 # The class used to provide CQL connections to Apache Cassandra.

--- a/docs/autogenerated/EccYamlFile.md
+++ b/docs/autogenerated/EccYamlFile.md
@@ -23,6 +23,7 @@
 **retryPolicy:**
 # Max number of attempts that ecChronos will try to connect with Cassandra.
 * maxAttempts: 5
+* unit: seconds
 # Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
 # If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
 * delay: 5

--- a/docs/autogenerated/EccYamlFile.md
+++ b/docs/autogenerated/EccYamlFile.md
@@ -13,24 +13,25 @@
 * host: localhost
 * port: 9042
 #
-# Connection Timeout for an CQL attempt.
+# Connection Timeout for a CQL attempt.
 # Specify a time to wait for cassandra to come up.
-# Connection is tried based on retry policy delay calculations. Each connection attempt will use timeout to calculate CQL Connection proccess delay.
+# Connection is tried based on retry policy delay calculations. Each connection attempt will use the timeout to calculate CQL connection process delay.
 #
 **timeout:**
 * time: 60
 * unit: seconds
 **retryPolicy:**
-# Max number of attempts that ecChronos will try to connect with Cassandra.
+# Max number of attempts ecChronos will try to connect with Cassandra.
 * maxAttempts: 5
-* unit: seconds
-# Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count.
-# If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay
+# Delay use to wait between an attempt and another, this value will be multiplied by the current attempt count powered by two.
+# If the current attempt is 4 and the default delay is 5 seconds, so ((4(attempt) x 2) x 5(default delay)) = 40 seconds.
+# If the calculated delay is greater than maxDelay, maxDelay will be used instead of the calculated delay.
 * delay: 5
-# The max delay that one attempt can wait until run.
+# Maximum delay before the next connection attempt is made.
 # Setting it as 0 will disable maxDelay and the delay interval will
 # be calculated based on the attempt count and the default delay.
 * maxDelay: 30
+* unit: seconds
 #
 # The class used to provide CQL connections to Apache Cassandra.
 # The default provider will be used unless another is specified.

